domain/application/state/charm_test.go

```diff
@@ -2344,6 +2349,33 @@ func (s *charmStateSuite) TestSetCharmThenGetCharmArchivePath(c *gc.C) {
 	c.Check(got, gc.DeepEquals, "archive")
 }
 
+func (s *charmStateSuite) TestSetCharmWithDuplicatedEndpointNames(c *gc.C) {
+	st := NewCharmState(s.TxnRunnerFactory())
+
+	_, err := st.SetCharm(context.Background(), charm.Charm{
+		Metadata: charm.Metadata{
+			Provides: map[string]charm.Relation{
+				"foo": {
+					Key:   "foo",
+					Name:  "baz",
+					Role:  charm.RoleProvider,
+					Scope: charm.ScopeGlobal,
+				},
+			},
+			Requires: map[string]charm.Relation{
+				"foo": {
+					Key:   "foo",
+					Name:  "baz",
+					Role:  charm.RoleProvider,
+					Scope: charm.ScopeGlobal,
+				},
+			},
+		},
+	}, setStateArgs("ubuntu"))
+
+	c.Assert(err, gc.ErrorMatches, `.*failed to insert charm relations.*`)
```

Comment:
This isn't ideal, I would prefer to check the type of the error, but there aren't really any that make sense here. The insert could fail for many reasons, so I suspect it isn't worth the complication

Reply:
So we should have a check in the service, that prevents this completely. This won't be redundant, but at least we've baked that into the tests.

Reply:
This verification has been added